### PR TITLE
cephadm: Modify the structure of the default container images

### PR DESCRIPTION
**Pull request by ShwetaBhosale1**
cephadm: Modify the structure of the default container images

Fixes: https://tracker.ceph.com/issues/68605

Signed-off-by: Shweta Bhosale <Shweta.Bhosale1@ibm.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ShwetaBhosale1** [comment]
This PR has cherry-picked commit of PR: https://github.com/ceph/ceph/pull/60377.

**Comment by adk3798** [comment]
I updated the PR title. It's not required to include the tracker being fixed in the PR title and having the component being updated in the title helps identify what's being changed without having to look through the changes.

File: src/cephadm/cephadmlib/container_types.py
Comment (adk3798) [inline src/cephadm/cephadmlib/container_types.py:12]: just to make you aware, I messed with these imports in the PR to move these to a common location (missed it originally as I didn't rebase the PR before testing) and those changes will conflict with these changes. You can either rebase your work on top of what I've since done with that PR, or wait until I merge it and solve conflicts. Up to you.

**Comment by ShwetaBhosale1** [comment]
jenkins test api

File: doc/man/8/cephadm.rst
Comment (adk3798) [inline doc/man/8/cephadm.rst:535]: We might want to add to this description that we list them in an ini format so the results can be modified and passed to the cluster with the `--config` flag during bootstrap.

File: src/pybind/mgr/cephadm/module.py
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:221]: These names like `container_image_prometheus` should ideally get populated from python-common/ceph/cephadm/images.py at this point. Maybe we should modify `DefaultImages` to only include the part of it specific to the daemon type (so `container_image_prometheus` for this one) and have callers handle prepending the `mgr/cephadm/` to the front. That will get us to the point where both the config name and the image are both coming from one spot. Currently you'd still have to specify the config option name both in images.py and here. Additionally, there is an idea to have these options, as in the entirety of
```
Option(
      'container_image_prometheus',
      default=DefaultImages.PROMETHEUS.value.name,
      desc='Prometheus container image',
)
```
be generated dynamically based on the contents of images.py so when you add a new image you only need to touch that one file. I don't think that portion needs to be in this PR, but what I was mentioning earlier about pulling `container_image_prometheus` from images.py rather than hardcoding it would be a step in that direction.

**Review by adk3798 — COMMENTED** [review]
What's here looks pretty good and works. Added a comment about further centralizing the config names on top of the images.

File: src/pybind/mgr/cephadm/module.py
Comment (ShwetaBhosale1) [reply src/pybind/mgr/cephadm/module.py]: Updated the PR

File: src/cephadm/cephadmlib/daemons/tracing.py
Comment (rkachach) [inline src/cephadm/cephadmlib/daemons/tracing.py:35]: to avoid this ".value.xxx" access we can add those properties to the `DefaultImages` class:

```
class DefaultImages(Enum):
    PROMETHEUS = ContainerImage(
        'quay.io/prometheus/prometheus:v2.51.0',
        'some_key_prometheus',
        'Prometheus container image'
    )

    @property
    def name(self):
        return self.value.name

    @property
    def key(self):
        return self.value.key

    @property
    def desc(self):
        return self.value.desc
```
This way, we can call `DefaultImages.JAEGER_COLLECTOR.name` directly

File: doc/cephadm/services/monitoring.rst
Comment (rkachach) [inline doc/cephadm/services/monitoring.rst:203]: I think this is really hard to maintain manually and since we are introducing a new class `DefaultImages` we should use sphinx autodoc for this. Plz, have a look to examples in other files such as mgmt-gateway.rst:

```
.. autoclass:: MgmtGatewaySpec
   :members:
```
In this case some customization is needed to generate the format we want but I think it should be feasible.

File: src/cephadm/cephadmlib/container_types.py
Comment (rkachach) [inline src/cephadm/cephadmlib/container_types.py:673]: This is equivalent to:

mgr_prefix = 'mgr/cephadm/'
mgr_images = {f"{mgr_prefix}{image.value.key}": image.value.name for image in DefaultImages}

File: src/python-common/ceph/cephadm/images.py
Comment (rkachach) [inline src/python-common/ceph/cephadm/images.py:8]: Technically this field should be named properly as it actually holds the image reference (i.e: `quay.io/prometheus/prometheus:v2.51.0` . I'd suggest/recommend using  `image_ref` instead of  just `name`.

**Review by rkachach — CHANGES_REQUESTED** [review]
This is a small detail but I think it's very important to name the field in a descriptive way.

**Review by adk3798 — APPROVED** [review]
Outside of comments Redouane has already provided, this LGTM.

File: src/cephadm/cephadmlib/daemons/tracing.py
Comment (ShwetaBhosale1) [reply src/cephadm/cephadmlib/daemons/tracing.py]: Thanks for pointing this! I have done the changes.

File: doc/cephadm/services/monitoring.rst
Comment (ShwetaBhosale1) [reply doc/cephadm/services/monitoring.rst]: Instead of adding here statically, i am creating a new rst file with images dynamically and including file here. Let me know if it looks good.

File: src/python-common/ceph/cephadm/images.py
Comment (ShwetaBhosale1) [reply src/python-common/ceph/cephadm/images.py]: done

File: doc/cephadm/services/monitoring.rst
Comment (ShwetaBhosale1) [reply doc/cephadm/services/monitoring.rst]: image

File: src/python-common/ceph/cephadm/images.py
Comment (adk3798) [inline src/python-common/ceph/cephadm/images.py:15]: looks like some formatting issues causing make check to fail
```
ceph/cephadm/images.py:8:19: E261 at least two spaces before inline comment
ceph/cephadm/images.py:8:20: E262 inline comment should start with '# '
ceph/cephadm/images.py:9:13: E261 at least two spaces before inline comment
ceph/cephadm/images.py:9:14: E262 inline comment should start with '# '
ceph/cephadm/images.py:10:14: E261 at least two spaces before inline comment
ceph/cephadm/images.py:15:1: E305 expected 2 blank lines after class or function definition, found 1
```

File: src/python-common/ceph/cephadm/images.py
Comment (rkachach) [inline src/python-common/ceph/cephadm/images.py:108]: @ShwetaBhosale1 all the introduced changes LGTM, thanks. I still see a room for improvement here as there's still a lot of repetition. I think we can still go beyond a little bit to simplify the Enum class definition with something like:


```
_img_prefix = 'container_image_'

def create_image(image_ref, description):
    key = description.lower().replace(' ', '_')
    return ContainerImage(
        image_ref,
        f'{_img_prefix}{key}',
        f'{description} container image'
    )

class MyDefaultImages(Enum):
    PROMETHEUS = create_image('quay.io/prometheus/prometheus:v2.51.0', 'Prometheus')
    LOKI = create_image('docker.io/grafana/loki:3.0.0', 'Loki')
    PROMTAIL = create_image('docker.io/grafana/promtail:3.0.0', 'Promtail')
    NODE_EXPORTER = create_image('quay.io/prometheus/node-exporter:v1.7.0', 'Node Exporter')
   .......
```
This way for a new container the developer has just to provide the `image_ref` and the description, but overall we:

- reduce repetition
- improve consistency (using the method there's no room for errors when generating the `ContainerImage`)
- improve code readability and Maintainability

**Review by rkachach — CHANGES_REQUESTED** [review]
Overall changes LGTM. I think there's still some room for improvement and we should be good to merge :)

File: doc/cephadm/services/monitoring.rst
Comment (rkachach) [reply doc/cephadm/services/monitoring.rst]: that looks really good @ShwetaBhosale1 thanks :+1:

File: src/python-common/ceph/cephadm/images.py
Comment (ShwetaBhosale1) [reply src/python-common/ceph/cephadm/images.py]: fixed this

File: src/python-common/ceph/cephadm/images.py
Comment (ShwetaBhosale1) [reply src/python-common/ceph/cephadm/images.py]: Updated the PR

**Review by rkachach — APPROVED** [review]
LGTM

File: src/pybind/mgr/cephadm/module.py
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/module.py:457]: :-D

File: src/python-common/ceph/cephadm/images.py
Comment (phlogistonjohn) [inline src/python-common/ceph/cephadm/images.py:16]: Tiny nitpick: if this function is not intended to be used outside of this file name it: `_create_image`. That way you can change it freely without worrying if code has imported and used this function elsewhere.

**Review by phlogistonjohn — APPROVED** [review]
Looks pretty good to me. I have one small nitpick that you can either follow up on in this PR or in a future one.

One other suggestion is to have the doc changes be a separate commit to the code change. It makes things easier on the reviewer and future explorers of the code.

File: src/python-common/ceph/cephadm/images.py
Comment (ShwetaBhosale1) [reply src/python-common/ceph/cephadm/images.py]: Updated the PR with function name change.

**Comment by ShwetaBhosale1** [comment]
jenkins test api

**Comment by ShwetaBhosale1** [comment]
jenkins test api

**Comment by ShwetaBhosale1** [comment]
jenkins test dashboard cephadm

**Comment by ShwetaBhosale1** [comment]
jenkins test api

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-11-19_02:02:13-orch:cephadm-wip-adk-testing-2024-11-18-1758-distro-default-smithi/

Failures:

- lots of upgrade failures, primarily because the dashboard cherrypy backports issue fix was still not present in reef when these tests ran
- 1 `MON_DOWN` cluster log warning, seen this before, need to update the ignoreslist for this test
- 1 selinux denial failure in an smb ctdb test, known issue
- 2 new mgmt-gateway tests being added by a PR in the run failed
- 1 strange failure applying an SMB spec. Seems like it could be a bug that is just rarely hit in that test, as no SMB related PRs were included in the run

I wouldn't merge any SMB changes or the new tests off of this run, but otherwise it's okay.

**Comment by ShwetaBhosale1** [comment]
Resolved merge conflict of nvmeof.py file

**Comment by ShwetaBhosale1** [comment]
jenkins test submodules

**Comment by ShwetaBhosale1** [comment]
jenkins test make check

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Comment by ShwetaBhosale1** [comment]
jenkins test make check

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Comment by ShwetaBhosale1** [comment]
jenkins test make check

**Comment by ShwetaBhosale1** [comment]
jenkins test make check

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Comment by ShwetaBhosale1** [comment]
jenkins test make check arm64

**Pull request merged**
